crates/core/c8y_api/src/smartrest/payload.rs

```diff
@@ -0,0 +1,98 @@
+use std::fmt::Display;
+
+use serde::Serialize;
+use tracing::error;
+
+use super::message::MAX_PAYLOAD_LIMIT_IN_BYTES;
+
+/// A Cumulocity SmartREST message payload.
+///
+/// A SmartREST message is either an HTTP request or an MQTT message that contains SmartREST topic and payload. The
+/// payload is a CSV-like format that is backed by templates, either static or registered by the user. This struct
+/// represents that payload, and should be used as such in SmartREST 1.0 and 2.0 message implementations.
+///
+/// # Example
+///
+/// ```text
+/// 503,c8y_Command,"This is a ""Set operation to SUCCESSFUL (503)"" message payload; it has a template id (503),
+/// operation fragment (c8y_Command), and optional parameters."
+/// ```
+///
+/// # Reference
+///
+/// - https://cumulocity.com/docs/smartrest/smartrest-introduction/
+#[derive(Debug, Clone, PartialEq, Eq)]
+// TODO: pub(crate) for now so it can be constructed manually in serializer::succeed_operation, need to figure out a
+// good API
```

Comment:
I wanted this type to eventually be used for all Smartrest messages, and for this it shouldn't make assumptions about which fields can be trimmed, as it's different between messages, e.g. it's safe to trim additional parameters in [506](https://cumulocity.com/docs/smartrest/mqtt-static-templates/#506), but it would be wrong to trim the last field, the URL, of [115](https://cumulocity.com/docs/smartrest/mqtt-static-templates/#115) message.

As it depends on the message which fields are optional and which are safe to trim, my view is that that trimming should be done by another layer and this `SmartrestPayload` type should only allow writing and reading valid payloads and minimize other assumptions.

Reply:
> As it depends on the message which fields are optional and which are safe to trim,

Good point. The caller has to tell which field can be trimmed.

> my view is that that trimming should be done by another layer and this SmartrestPayload type should only allow writing and reading valid payloads and minimize other assumptions.

Okay for now, but I'm not really convinced with idea for yet another layer. We already have  too many layers.